### DISCUSSION TITLE
Retiring @Table

### DISCUSSION THREAD

**Discussion by gavinking** [hibernate/hibernate-orm#5359] [Design Proposals]
The `@Table` annotation has been sort of semi-deprecated for a long time now, but it has a zombie second life as a way to specify certain things for secondary tables.

I would really like to fully-deprecate this thing, and eventually remove it, but I'm having trouble coming up with something elegant to replace it with.

Currently, its useful bits are:

- `checkConstraint` and `comment`
- `fetch`, `inverse`, and `optional`
- `sqlInsert`, `sqlUpdate`, and `sqlDelete`

and they're all _only_ needed for secondary tables.

Now, I'm [hoping](https://github.com/jakartaee/persistence/issues/381) to see JPA add `check` and `comment` directly to the JPA `@Table` and `@SecondaryTable` annotations but obviously that will take a while.

But the other things don't make sense for the standard.

Honestly the best thing I've come up with so far is to simply move these things to an `@Secondary` annotation:

```java
public @interface Secondary {
	SecondaryTable table(); // maybe, or perhaps we just match them by order?

	Check check() default @ Check("");
	Comment comment() default @Comment("");

	FetchMode fetch() default FetchMode.JOIN;
	boolean inverse() default false;
	boolean optional() default true;

	SQLInsert sqlInsert() default @SQLInsert(sql="");
	SQLUpdate sqlUpdate() default @SQLUpdate(sql="");
	SQLDelete sqlDelete() default @SQLDelete(sql="");
}
```

but this really isn't all that different from `@Table` as it exists today. (Though the _name_ is a lot better.)

**Comment by beikov** [comment]
Sounds like a good idea. I guess I'd prefer calling it `SecondaryTableOptions` or something like that though, and also have an optional `String table() default ""` which allows to match by index or name instead of `SecondaryTable table()`.

**Reply by yrodiere** [reply]
I confirm that, at least from my outside perspective, `@Secondary` is a bit too generic and not very clear.

**Comment by gavinking** [comment]
Curiously, `@Table(fetch=SELECT)` does nothing in H6, and the guts of the underlying implementation was completely removed. (Though there was some stuff left lying around, and the annotation member was not deprecated!)

I've cleaned up the zombie code and deprecated this member.

**Comment by gavinking** [comment]
> but this really isn't all that different from `@Table` as it exists today

Upon reflection, I don't want to add a big "container" annotation, so I would rather split this stuff out into smaller annotations. I'm going to begin by adding `@SecondaryRow` for the `inverse` and `optional` members.

**Reply by gavinking** [reply]
And I think I'll just add a `table` member to `@SqlInsert` and friends. It's stringly-typed, which I hate, but that's the case for the whole `@SecondaryTable` concept, so I think I can live with it here.

**Reply by gavinking** [reply]
That leaves `@Check` and `@Comment`, which I'm still not sure what to do with, unless JPA accepts my proposal for adding these directly to `@SecondaryTable`.

**Comment by beikov** [comment]
Seems to have been completed already as part of https://hibernate.atlassian.net/browse/HHH-15570, so I'll close this discussion.

**Reply by gavinking** [reply]
It's not really finished:

> That leaves `@Check` and `@Comment`, which I'm still not sure what to do with, unless JPA accepts my proposal for adding these directly to `@SecondaryTable`.

**Reply by gavinking** [reply]
JPA has integrated my proposal, so this work is now finished.
